File: kiosk/src/lib/enrollGate.ts
/**
 * What RFID Enroll refuses to send, and why.
 *
 * The tag box accepts whatever the reader types, and a reader is a
 * keyboard: the two ways an operator ruins an asset record here are both
 * a second scan of something that was never a tag.
 *
 *   1. **The asset scanned twice.** The barcode on the rack is right
 *      there, and scanning it again while the tag box has focus would
 *      write an asset ID into `rfid_tag` — a value the portal would
 *      accept (it is alphanumeric and short enough) and nobody would
 *      notice until a reader failed to find the asset. So an entry that
 *      matches any asset ID or serial in the synced roster is refused,
 *      naming what it actually is.
 *   2. **A tag that already belongs to something.** `assets_rfid_uniq`
 *      makes this the portal's call, and the endpoint answers 409
 *      `rfid_in_use` — but the common case is the tag enrolled seconds
 *      ago on the previous asset, and a round trip to be told so is
 *      slower than the operator's next move. The roster and this
 *      session's log (`enrollLog`) both get asked first.
 *
 * Neither check replaces the server's. This is the fast, local half; the
 * save still runs, still gets 409'd on anything this could not know, and
 * feeds what it learns back into the log.
 *
 * The order matters: the roster is consulted before the format, so
 * scanning a hyphenated serial says "that's the serial for X" rather
 * than complaining about the hyphen.
 */

import { tagHolder } from './enrollLog';
import { padRfid, rfidProblemText } from './rfid';
import {
  matchAssetOrSerial, matchScan, type ScanAsset, type ScanIndex, type ScanMatchKind,
} from './scanMatch';

export type GateResult =
  | { ok: true; tag: string }
  | { ok: false; message: string };

/** How an asset is named in a refusal — assets can reach the kiosk
 *  without a name, and "That's the asset ID for null" helps nobody. */
function nameOf(asset: ScanAsset): string {
  return asset.name || asset.asset_id || 'another asset';
}

function kindWord(kind: ScanMatchKind): string {
  return kind === 'asset_id' ? 'asset ID' : 'serial';
}

/** Decide whether `raw` may be written onto `asset` as its RFID tag.
 *  On success the tag comes back in the stored 24-character form. */
export function checkTagEntry(
  index: ScanIndex, asset: ScanAsset, raw: string,
): GateResult {
  // 1. Is this the asset's own barcode, or some other asset's?
  const asAsset = matchAssetOrSerial(index, raw);
  if (asAsset) {
    const word = kindWord(asAsset.kind);
    return {
      ok: false,
      message: asAsset.asset.id === asset.id
        ? `That's this asset's own ${word}, not an RFID tag.`
        : `That's the ${word} for ${nameOf(asAsset.asset)}, not an RFID tag.`,
    };
  }

  // 2. Is it a tag at all?
  const { tag, problem } = padRfid(raw);
  if (problem) return { ok: false, message: rfidProblemText(problem) };

  // 3. Does something else already hold it? The session's own log first
  //    — it knows about saves the roster has not been re-synced for.
  const logged = tagHolder(tag!);
  if (logged && logged.assetId !== asset.id) {
    return { ok: false, message: `That tag is already on ${logged.assetName}.` };
  }
  const onRoster = matchScan(index, tag!);
  if (onRoster?.kind === 'rfid' && onRoster.asset.id !== asset.id) {
    return { ok: false, message: `That tag is already on ${nameOf(onRoster.asset)}.` };
  }

  return { ok: true, tag: tag! };
}

File: Android_Kiosk_App/app/src/main/java/com/serversherpa/kiosk/core/scan/EnrollGate.kt
package com.serversherpa.kiosk.core.scan

/**
 * The gates RFID Enroll puts in front of a save.
 *
 * Two things go wrong on a busy floor: the same asset gets walked past the
 * reader twice, and the same tag gets waved at two assets. Both are caught
 * here, from what this kiosk already knows, so the operator hears about it
 * at the box instead of after a round trip. The portal still has the last
 * word — these checks narrow what reaches it, they do not replace it.
 */

/** One line of this session's enroll log: what this kiosk has already put on
 *  an asset since the screen opened. Memory only — it is a gate against a
 *  repeat scan, not a record. The portal holds the record. */
data class EnrollLogEntry(val assetRowId: String, val tag: String, val assetName: String)

enum class EnrollTagIssue {
    /** The tag being scanned is the one already on this asset. */
    SAME_TAG_ON_THIS_ASSET,
    /** This kiosk put the tag on something else a moment ago. */
    TAG_USED_THIS_SESSION,
    /** The synced roster says the tag belongs to another asset. */
    TAG_ON_ANOTHER_ASSET,
}

data class EnrollTagVerdict(val issue: EnrollTagIssue, val holder: String?)

fun enrollTagText(verdict: EnrollTagVerdict): String {
    val holder = verdict.holder ?: "another asset"
    return when (verdict.issue) {
        EnrollTagIssue.SAME_TAG_ON_THIS_ASSET -> "That tag is already on this asset."
        EnrollTagIssue.TAG_USED_THIS_SESSION -> "You just enrolled that tag on $holder. Scan a different tag."
        EnrollTagIssue.TAG_ON_ANOTHER_ASSET -> "That tag is on $holder. Scan a different tag."
    }
}

/** What this session already did to an asset, if anything. */
fun enrolledThisSession(log: List<EnrollLogEntry>, assetRowId: String): EnrollLogEntry? =
    log.firstOrNull { it.assetRowId == assetRowId }

/**
 * Why this tag must not be saved onto this asset, or null to go ahead.
 *
 * The session log is consulted before the roster: it is the fresher of the
 * two, and after a save that the local roster update missed it is the only
 * one that knows.
 */
fun <A : ScanAsset> checkEnrollTag(
    index: ScanIndex<A>?,
    log: List<EnrollLogEntry>,
    target: A,
    padded: String,
): EnrollTagVerdict? {
    val want = rfidKey(padded) ?: return null
    if (rfidKey(target.rfid) == want) return EnrollTagVerdict(EnrollTagIssue.SAME_TAG_ON_THIS_ASSET, null)
    log.firstOrNull { rfidKey(it.tag) == want && it.assetRowId != target.id }
        ?.let { return EnrollTagVerdict(EnrollTagIssue.TAG_USED_THIS_SESSION, it.assetName) }
    index?.byRfid?.get(want)?.takeIf { it.id != target.id }
        ?.let { return EnrollTagVerdict(EnrollTagIssue.TAG_ON_ANOTHER_ASSET, it.name ?: it.assetId) }
    return null
}
